Enhance docker image

kaiii708 opened the pull request:
This commit updates the Docker image to include the following executables:
- rv32emu-user: for user-mode emulation
- rv32emu-system: for system-mode emulation
- rv_histogram: for statistical analysis

To keep the image compact, build artifacts are excluded from the final image. Instructions for retrieving prebuilt binaries are added to the README.

Close #581

visitorckw commented on Dockerfile:
```diff
@@ -9,17 +9,22 @@ COPY . .
+RUN make ENABLE_SYSTEM=1 INITRD_SIZE=32
 
-FROM alpine:3.19 AS final
+FROM alpine:3.21 AS final
```
Which parts of this change rely on Alpine 3.21 and wouldn't work on 3.19?

kaiii708 replied:
Commit 37c9838 mentioned that the end of support for Alpine Linux v3.19.x is on 1 Nov 2025, so I upgraded it to 3.21.

visitorckw commented:
This patch updates the Dockerfile but also makes large README changes.
I’d prefer splitting them into separate commits so each does one thing and review is easier.

visitorckw commented:
Also, I'd prefer using the imperative mood when describing changes [1].
E.g. This commit updates ... -> Update ...

Link: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/submitting-patches.rst?h=v6.17-rc3#n94 [1]

kaiii708 committed:
Bump Alpine Linux in Docker final stage

Commit 37c9838 ("Bump Alpine Linux for Docker image") updated
only the build stage but left the final stage unchanged. This commit
completes the version bump by upgrading the final stage as well.

kaiii708 committed:
Enhance Docker image with rv32emu-user, -system, and -histogram binaries

Include three executables in the Docker image:
- rv32emu-user: for user-mode emulation
- rv32emu-system: for system-mode emulation
- rv32emu-histogram: for statistical analysis

This enhancement enables the Docker image to support both user-mode
and system-mode emulation.

Close #581

kaiii708 committed:
Update README for Docker image artifacts

Document how to fetch and run user-mode and system-mode artifacts
from the rv32emu-prebuilt repository when using the Docker image.

Link: https://github.com/sysprog21/rv32emu-prebuilt

kaiii708 commented:
Split into 3 commits. Please take another look. Thanks!

kaiii708 commented:
The CI build has succeeded, and the Docker image is available here:  
https://github.com/kaiii708/rv32emu/actions/runs/17394344573  
Test it with:  
```bash
docker run --rm -it kaiii708/rv32emu
```

visitorckw commented on Dockerfile:
```diff
@@ -9,17 +9,22 @@ COPY . .
 RUN make
 RUN make tool
+RUN mv ./build/rv32emu ./build/rv32emu-user
+RUN mv ./build/rv_histogram ./build/tmp_rv_histogram
```
Renaming rv_histogram to tmp_rv_histogram and then back looks a bit hacky and may confuse readers.
Since we already renamed rv32emu to rv32emu-user/rv32emu-system, how about also giving rv_histogram an rv32emu- prefix?
That way all three binaries can be renamed to their final names right after build, and the later COPY steps become cleaner and easier to follow.

visitorckw commented on README.md:
````diff
@@ -143,6 +143,40 @@ The usage and limitations of Doom and Quake demo are listed in [docs/demo.md](do
+  -k rv32emu-linux-image-prebuilt/linux-image/Image \
+  -i rv32emu-linux-image-prebuilt/linux-image/rootfs.cpio
+```
+
````
Since we renamed rv_histogram, could we also update the README's Docker section to document how to run the histogram tool in addition to the user/system mode emulators?

jserv commented:
Thank @kaiii708 for contributing!
